### DISCUSSION TITLE
Create an attribute that, when applied to a struct type, indicates that the default value of that struct is invalid

### DISCUSSION THREAD

**Discussion by Xenoprimate** [dotnet/csharplang#7632] [Language Ideas]
With the latest versions of C# it is possible to override the default ctor for struct types which is useful to help prevent badly-formed struct-type values, but does not prevent usage of the default(T) value of a struct-type (either explicity or implicitly e.g. as part of an array or non-initialized class member).

Although the compiler can never fully prevent someone from creating a value of default(T), having an attribute we can apply to struct types that gives a hint would be useful. In these cases, where the compiler detects that a value will be default(T) for a type T that is annotated with this proposed attribute, it can emit a warning.

The compiler would only have to detect the most obvious cases-- but these would still be very useful in minimizing errors.

Examples always help, so here's an example:

```
[DefaultValueIsInvalid] // Proposed attribute
readonly struct UnitQuaternion {
    public float X;
    public float Y;
    public float Z;
    public float W;
    // etc...
}

// Then, later:

static void Test() {
    var uqArr = new UnitQuaternion[4]; // Warning here unless we Array.Fill?
    DoSomething(uqArr[0]); // Or warning here that we're probably using an invalid value?
}

class SomeClass {
    UnitQuaternion _uq; // Warning here that _uq will be left at default(T)?
}
```

**Comment by Xenoprimate** [comment]
Closing, I think this should have been in the roslyn repo actually. Will recreate it there.

**Reply by 333fred** [reply]
This should not be in the roslyn repo, it should be here. However, it is a duplicate of a lot of other discussions on csharplang, such as https://github.com/dotnet/csharplang/issues/146, https://github.com/dotnet/csharplang/blob/main/meetings/2020/LDM-2020-07-01.md#non-defaultable-struct-types, https://github.com/dotnet/csharplang/issues/1981, and https://github.com/dotnet/csharplang/issues/2019.
